Sample profiling

adrosenbaum opened the pull request:
Implementation of sample profiling.

User must first load 'profile variants' to the profile_variant collection
    loqusdb profile <vcf_file>

To activate the profiling function when using the load subcommand give the --check-profiles flag and --profile-threshold option
    loqusdb load ... --check-profiles --profile-theshold <float 0-1>

This will check the genotype of all samples in a vcf-file (not sv), at the positions of the profile_variants in the 'profile_variant' collection. The profile is assembled as a string for each sample. The profiles are checked against all profiles of previous samples in the database. If the check passes, the case and its variants are loaded. The profile string is found in the 'case' collection for each individual.

todo in this PR:
    tests

adrosenbaum committed:
Added support to load variants used for profiling

-New collection 'profile_variant'
-New CLI subcommand 'profile', takes a vcf as input. The variants in the
vcf will populate the 'profile_variant' collection, to be used for
profiling of samples.

adrosenbaum committed:
Implementation of sample profiling

This is a first implementation of sample profiling. For each new case,
with a snv vcf given, a profile for each sample will be made, that is
stored in the case collection, for each individual. Upon uploading new
cases, the profiles of the loaded individuals will be compared against
all profiles in the database. If there already exists similar profiles
the loading will be aborted.

adrosenbaum committed:
Add --check-profile flag for CLI

User can choose to use the sample profiling with flag --check-profile,
i.e. loqusdb load ... --check-profile.
User can specify the threshold of similarity with --profile-threshold

moonso commented:
Hi @adrosenbaum , nice work! I'll wait for the tests and start with the review then. It makes it easier to see what is going on.

adrosenbaum committed:
Added some tests

moonso commented on loqusdb/build_models/profile_variant.py:
```diff
@@ -0,0 +1,55 @@
+def get_maf(variant):
+    """
+        if ID CAF exists in INFO column, return the allele frequency for
+        the alt allele
```
Could you follow the `Args: Returns:` syntax? Also can you elaborate a little bit more. What type is returned?

moonso commented on loqusdb/build_models/profile_variant.py:
```diff
@@ -0,0 +1,55 @@
+        if ID CAF exists in INFO column, return the allele frequency for
+        the alt allele
+    """
+
```
To make it more flat:
```python
if not variant.INFO.get('CAF'):
    return None
maf_list = json.loads(variant.INFO.get('CAF'))
return maf_list[1]
```

moonso commented on loqusdb/commands/load.py:
```diff
@@ -14,6 +14,12 @@
+def validate_profile_threshold(ctx, param, value):
+    if not (0 <= value <= 1):
+        raise ValueError('threshold must be between 0-1')
+    else:
```
```suggestion

```

moonso commented on loqusdb/commands/load.py:
```diff
@@ -14,6 +14,12 @@
+    if not (0 <= value <= 1):
+        raise ValueError('threshold must be between 0-1')
+    else:
+        return value
```
```suggestion
    return value
```

moonso commented on loqusdb/commands/load_profile.py:
```diff
@@ -0,0 +1,18 @@
+@click.argument('vcf_file', type = click.Path(exists=True))
+@click.pass_context
+def load_profile(ctx, vcf_file):
+
```
Please add a doctring with some more information like:
```python
"""
Takes a vcf with variants choosen for profile

Profiling is used when...
....
...
"""
```

moonso commented on loqusdb/plugins/mongo/adapter.py:
```diff
@@ -4,14 +4,14 @@
 from mongo_adapter import MongoAdapter as BaseAdapter
 
-from . import VariantMixin, CaseMixin
+from . import VariantMixin, CaseMixin, ProfileVariantMixin
```
```suggestion
from . import (VariantMixin, CaseMixin, ProfileVariantMixin)
```

moonso commented on loqusdb/plugins/mongo/profile_variant.py:
```diff
@@ -0,0 +1,28 @@
+
+    def add_profile_variants(self, profile_variants):
+
+        """Add several variants to database
```
```suggestion
        """Add several variants to the profile collection in the database
```

moonso commented on loqusdb/utils/profiling.py:
```diff
@@ -0,0 +1,137 @@
+    """
+        given a dict of profiles, searches through all the samples in the DB
+        for a match.
+
```
Could you write some more about what is happening when a match is found?

moonso reviewed (changes requested):
Nice work @adrosenbaum !! I like the testing as well.
There where some comments, minor stuff. Could you add some stuff to the documentation? It needs to be explained somewhere why this is here and how it works

adrosenbaum committed:
Added documentation, update, and stats options for profiles.

Documentation of new profiling feature in docs/user-guide/profiling.md.
More comments to code.

Added feature 'loqusdb profile --update' will update the profile for all
existing samples in the database.

Added feature 'loqusdb profile --stats' prints some basic statistics
from comparing all the profiles in the database.

adrosenbaum commented:
Fixed what you commented on @moonso. 

I've also added a functionality that will update the profiles of existing samples in the database, if a path to a snv VCF file is given for the case. (loqusdb profile --update).

There is also a feature 'loqusdb profile --stats' that I thought should print some statistics based on the profiles. I am however unsure of what statistics might be useful to look at (or if this feature is necessary at all). Now it calculates the pair wise hamming distance between all sample profiles in the DB, and calculates the mean, and std, to get an idea of how well the samples can be distinguished from each other using this method.

adrosenbaum committed:
Change in stats feature

'loqusdb profile --stats' now prints a table, showing the number of
hamming distances found in some ranges.

'loqusdb profile --stats --profile-threshold <float>' sets a
similarity threshold, for all hamming distances found above this, this
issues a warning stating what samples are similar to each other.

adrosenbaum committed:
added option --soft-threshold to load subcommand

Usert can now also provide a soft threshold, e.g.
'loqusdb load ... --hard-threshold 0.95 --soft-threshold 0.90'

If any sample profile has a hamming distance above --hard-threshold to
any other sample in the database, the load will be aborted.

If any sample profile has a hamming distance above --soft-threshold, the
load will continue, but the matching samples found in the database will
be stored for that individual in the database.

adrosenbaum commented:
Pushed some new commits.

Do you have the possibility to review this soon @moonso? It would be nice to test this branch tomorrow on stage.

adrosenbaum committed:
Change from CAF to MAF tag when looking for MAF in vcf file

adrosenbaum committed:
fix small bug

adrosenbaum committed:
compare genotypes, instead of alleles in profile string

adrosenbaum committed:
The vcf used for sample profiling is gived seperately from variants to
be loaded.

the --check-profile option now takes the vcf file from the case that is
used for profiling. Before the profiling was done on the --variant-file
vcf.

adrosenbaum commented:
Hopefully final change.

When loading a case, the vcf that the profiling is made on is entered separate from the variants that is to be loaded into the db. The 'profile' vcf is given with the --check-profile option. Any thoughts @moonso and @henrikstranneheim?

henrikstranneheim commented:
@moonso Do you have time to look at this PR today?

moonso commented:
Ok I'll have a look sometime today!

henrikstranneheim commented:
👍

moonso reviewed (approved):
Looks great @adrosenbaum ! All I would like now are a couple of lines in the documentation about the idea of the profiling feature. Why it is there and why one needs to point to another VCF than the load vcf. This might be clear to us but if another user is uploading it would be nice to have it in the docs. 👍

moonso commented:
Oh and can you update the CHANGELOG.md?

adrosenbaum committed:
Added to CANGELOG.md and documentation

adrosenbaum commented:
Added to documentation and CHANGELOG.md. Does this sufficiently describe the new profiling feature @moonso? If so I'll merge with master.

moonso commented:
Great! Go ahead
